runtimebp/cpu.go

```diff
@@ -12,75 +12,120 @@ import (
 func NumCPU() float64 {
+	// Big enough buffer to read the numbers in the files wholly into memory.
+	buf := make([]byte, 1024)
+
+	n, err := numCPUCgroupsV2(buf)
+	if err == nil {
+		return n
+	}
+
+	// fallback to cgroups v1
+	fmt.Fprintf(
+		os.Stderr,
+		"NumCPU: Failed to read cgroup v2, fallback to cgroup v1: %v\n",
+		err,
+	)
+	n, err = numCPUCgroupsV1(buf)
+	if err == nil {
+		return n
+	}
+
+	// Fallback to the standard runtime package value.
+	fmt.Fprintf(
+		os.Stderr,
+		"NumCPU: Failed to read cgroup v1, fallback to NumCPU on the physical machine: %v\n",
+		err,
+	)
+	return float64(runtime.NumCPU())
+}
+
+func numCPUCgroupsV2(buf []byte) (float64, error) {
 	const (
-		quotaPath  = "/sys/fs/cgroup/cpu/cpu.cfs_quota_us"
-		periodPath = "/sys/fs/cgroup/cpu/cpu.cfs_period_us"
+		maxPath = "/sys/fs/cgroup/cpu.max"
 	)
 
-	// Default to the standard runtime package value.
-	defaultCPUs := float64(runtime.NumCPU())
+	values, err := readNumbersFromFile(maxPath, buf)
+	if err != nil || len(values) != 2 {
+		return 0, fmt.Errorf(
+			"runtimebp: failed to read max file, falling back to use runtime.NumCPU(): %w, %v",
+			err,
+			values,
+		)
+	}
 
-	// Big enough buffer to read the number in the file wholly into memory.
-	buf := make([]byte, 1024)
+	return values[0] / values[1], nil
+}
 
-	quota, err := readNumberFromFile(quotaPath, buf)
-	if err != nil {
-		fmt.Fprintf(
-			os.Stderr,
-			"NumCPU: Failed to read quota file, falling back to use runtime.NumCPU(): %v\n",
+func numCPUCgroupsV1(buf []byte) (float64, error) {
+	const (
+		quotaPath  = "/sys/fs/cgroup/cpu/cpu.cfs_quota_us"
+		periodPath = "/sys/fs/cgroup/cpu/cpu.cfs_period_us"
+	)
+
+	quota, err := readNumbersFromFile(quotaPath, buf)
+	if err != nil || len(quota) != 1 {
+		return 0, fmt.Errorf(
+			"runtimebp: failed to read quota file, falling back to use runtime.NumCPU(): %w, %v",
 			err,
+			quota,
 		)
-		return defaultCPUs
 	}
 
 	// CFS quota returns -1 if there is no limit, return the default.
-	if quota < 0 {
-		fmt.Fprintf(
-			os.Stderr,
-			"NumCPU: Quota file returned %f, falling back to use runtime.NumCPU(): %v\n",
-			quota,
+	if quota[0] < 0 {
+		return 0, fmt.Errorf(
+			"runtimebp: quota file returned %f, falling back to use runtime.NumCPU(): %v",
+			quota[0],
 			err,
 		)
-		return defaultCPUs
 	}
 
-	period, err := readNumberFromFile(periodPath, buf)
-	if err != nil {
-		fmt.Fprintf(
-			os.Stderr,
-			"NumCPU: Failed to read period file, falling back to use runtime.NumCPU(): %v\n",
+	period, err := readNumbersFromFile(periodPath, buf)
+	if err != nil || len(period) != 1 {
+		return 0, fmt.Errorf(
+			"runtimebp: failed to read period file, falling back to use runtime.NumCPU(): %w, %v",
 			err,
+			period,
 		)
-		return defaultCPUs
 	}
 
-	return quota / period
+	return quota[0] / period[0], nil
 }
 
-func readNumberFromFile(path string, buf []byte) (float64, error) {
+func readNumbersFromFile(path string, buf []byte) ([]float64, error) {
```

Comment:
the problem with `os.ReadFile` is that there's no control of how many allocations it could make. I changed to use `io.ReadFull` (and handle `io.ErrUnexpectedEOF`) instead.

Reply:
This is only called once at binary startup, right? Who cares how many allocations it does :)

Reply:
you can't stop people from calling `runtimebp.NumCPU` from non-startup code (currently `runtime.NumCPU` is also not limited to be called only at startup code either) :)

also it's just my habit of defensive coding to be wary of things like `os.ReadFile` that could return any arbitrary large size of `[]byte` when used on the wrong file.